test/Fencer/Rules/Test.hs

```diff
@@ -227,3 +230,57 @@ test_rulesLoadRulesReadPermissions =
   file2 = simpleRuleFile
     ("domain2" </> "config" </> "config.yml")
     domainDescriptorKeyText
+
+-- | test that faulty rules in reloading with 'loadRulesFromDirectory'
+-- are rejected and the valid ones that were previously loaded are
+-- kept instead.
+--
+-- This matches the behavior of @lyft/ratelimit@.
+test_rulesReloadRules :: TestTree
+test_rulesReloadRules =
+  testCase "Rules fail to reload for an invalid domain" $
+    Temp.withSystemTempDirectory "fencer-config" $ \tempDir -> do
+      rules <- writeAndLoadRules
+        (#ignoreDotFiles False)
+        (#root tempDir)
+        (#files files)
+      failures <- writeAndLoadRules
+        (#ignoreDotFiles False)
+        (#root tempDir)
+        (#files files')
+      case (rules, failures) of
+        (Left errs, _) ->
+          assertFailure
+            ("Expected domains, got failures: " ++
+             prettyPrintErrors (NE.toList errs))
+        (Right _, Right newDefinitions) ->
+          assertFailure
+            ("Expected failures, got domains: " ++ show newDefinitions)
+        (Right definitions, f@(Left _)) -> do
+          assertEqual
+            "unexpected failure"
+            (length . toErrorList $ mapLeft NE.toList f)
+            (length . toErrorList $ expectedFailure)
```

Comment:
This is a bit weird. So you have a particular failure

```
"Error in $.descriptors[1]: key \"key\" not present"
```

but then you just remove all the details and only use the fact that there's supposed to be only a single failure. I'd say that you either should write

```haskell
          assertEqual
            "unexpected failure"
            (length . toErrorList $ mapLeft NE.toList f)
            1
```

or compare actual errors, not just the numbers of them.

Reply:
I changed this to comparing errors instead of their count.